File: MainWindow.py
# ssh key test
import os

from PySide2.QtCore import QCoreApplication, QSettings, QByteArray, QFileSystemWatcher
from PySide2.QtWidgets import QMainWindow, QVBoxLayout, QWidget, QSpinBox, QLabel, QSizePolicy, QHBoxLayout, \
    QPushButton, QMessageBox, QFileDialog, QCheckBox
from PySide2.QtGui import QIcon, QPixmap, QPainter, QColor, QFontMetrics, QFont, QPen, QImageReader

from ColorEdit import ColorEdit
from Generator import Generator
from PixmapWidget import PixmapWidget

class MainWindow(QMainWindow):

    def __init__(self, app, parent=None):
        super(MainWindow, self).__init__(parent)
        self.imagesDir = app.dir + '/images/'
        self.setWindowIcon(QIcon(self.imagesDir + 'icon.png'))
        self.path = ''

        self.settings = QSettings()
        self.lastDir = self.settings.value('lastDir', '')

        self.setMinimumWidth(540)

        self.supportedFormats = []
        for f in QImageReader.supportedImageFormats():
            self.supportedFormats.append(str(f.data(), encoding="utf-8"))

        self.fileWatcher = QFileSystemWatcher()
        self.fileWatcher.fileChanged.connect(self.fileChanged)

        # widgets
        self.showPixmapWidget = None

        self.tileWidthSpinBox = QSpinBox()
        self.tileWidthSpinBox.setValue(16)
        self.tileWidthSpinBox.setFixedWidth(50)
        self.tileWidthSpinBox.setMinimum(1)

        self.tileHeightSpinBox = QSpinBox()
        self.tileHeightSpinBox.setValue(16)
        self.tileHeightSpinBox.setFixedWidth(50)
        self.tileHeightSpinBox.setMinimum(1)

        self.paddingSpinBox = QSpinBox()
        self.paddingSpinBox.setFixedWidth(50)
        self.paddingSpinBox.setMinimum(1)

        self.transparentCheckbox = QCheckBox("Transparent")
        self.transparentCheckbox.setChecked(True)
        self.transparentCheckbox.stateChanged.connect(self.transparentChanged)

        self.backgroundColorEdit = ColorEdit()
        self.backgroundColorEdit.setEnabled(False)
        self.backgroundColorLabel = QLabel("Background color:")
        self.backgroundColorLabel.setEnabled(False)

        self.forcePotCheckBox = QCheckBox("Force PoT")
        self.forcePotCheckBox.setChecked(True)
        self.forcePotCheckBox.stateChanged.connect(self.forcePotChanged)

        self.reorderTilesCheckBox = QCheckBox("Reorder tiles")

        self.generateAndExportButton = QPushButton("Generate and export")
        self.generateAndExportButton.setFixedHeight(32)
        self.generateAndExportButton.clicked.connect(self.generateAndExportClicked)
        self.generateAndExportButton.setEnabled(False)

        self.pixmapWidget = PixmapWidget()
        self.pixmapWidget.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding)
        self.pixmapWidget.setPixmap(self.createDropTextPixmap())
        self.pixmapWidget.dropSignal.connect(self.fileDropped)
        self.pixmapWidget.setMinimumHeight(300)

        # load settings
        self.tileWidthSpinBox.setValue(int(self.settings.value('tileWidth', 16)))
        self.tileHeightSpinBox.setValue(int(self.settings.value('tileHeight', 16)))
        self.paddingSpinBox.setValue(int(self.settings.value('padding', 1)))
        self.forcePotCheckBox.setChecked(True if self.settings.value('forcePot', 'true') == 'true' else False)
        self.reorderTilesCheckBox.setChecked(True if self.settings.value('reorderTiles', 'false') == 'true' else False)
        self.transparentCheckbox.setChecked(True if self.settings.value('transparent', 'false') == 'true' else False)
        self.backgroundColorEdit.setColorText(str(self.settings.value('backgroundColor', '#FF00FF')))
        self.restoreGeometry(QByteArray(self.settings.value('MainWindow/geometry')))
        self.restoreState(QByteArray(self.settings.value('MainWindow/windowState')))

        # layout
        hl1 = QHBoxLayout()
        hl1.setContentsMargins(5, 5, 5, 5)
        hl1.addWidget(QLabel("Tile width:"))
        hl1.addSpacing(5)
        hl1.addWidget(self.tileWidthSpinBox)
        hl1.addSpacing(15)
        hl1.addWidget(QLabel("Tile height:"))
        hl1.addSpacing(5)
        hl1.addWidget(self.tileHeightSpinBox)
        hl1.addSpacing(15)
        hl1.addWidget(QLabel("Padding:"))
        hl1.addSpacing(5)
        hl1.addWidget(self.paddingSpinBox)
        hl1.addSpacing(15)
        hl1.addWidget(self.forcePotCheckBox)
        hl1.addSpacing(15)
        hl1.addWidget(self.reorderTilesCheckBox)
        hl1.addStretch()

        hl2 = QHBoxLayout()
        hl2.setContentsMargins(5, 5, 5, 5)
        hl2.addWidget(self.transparentCheckbox)
        hl2.addSpacing(15)
        hl2.addWidget(self.backgroundColorLabel)
        hl2.addSpacing(5)
        hl2.addWidget(self.backgroundColorEdit)
        hl2.addStretch()

        hl3 = QHBoxLayout()
        hl3.setContentsMargins(5, 5, 5, 5)
        hl3.addWidget(self.generateAndExportButton)

        vl = QVBoxLayout()
        vl.setContentsMargins(0, 0, 0, 0)
        vl.setSpacing(0)
        vl.addLayout(hl1)
        vl.addLayout(hl2)
        vl.addWidget(self.pixmapWidget)
        vl.addLayout(hl3)

        w = QWidget()
        w.setLayout(vl)
        self.setCentralWidget(w)

        self.setTitle()

    def setTitle(self):
        p = ' - ' + os.path.basename(self.path) if self.path else ''
        self.setWindowTitle(QCoreApplication.applicationName() + ' ' + QCoreApplication.applicationVersion() + p)

    def createDropTextPixmap(self):
        pixmap = QPixmap(481, 300)
        pixmap.fill(QColor("#333333"))
        painter = QPainter(pixmap)
        font = QFont("Arial")
        font.setPixelSize(28)
        font.setBold(True)
        fm = QFontMetrics(font)
        painter.setFont(font)
        painter.setPen(QPen(QColor("#888888"), 1))
        text = "Drop the tileset image here"
        x = (pixmap.width()-fm.width(text))/2
        y = (pixmap.height()+fm.height())/2
        painter.drawText(x, y, text)
        del painter
        return pixmap

    def fileDropped(self, path):
        path = str(path)
        name, ext = os.path.splitext(path)
        ext = ext[1:]
        if not ext in self.supportedFormats:
            QMessageBox.warning(self, "Warning", "The dropped file is not supported")
            return
        pixmap = QPixmap(path)
        if pixmap.isNull():
            QMessageBox.warning(self, "Warning", "Can't load the image")
            return
        if self.path:
            self.fileWatcher.removePath(self.path)
        self.path = path
        self.fileWatcher.addPath(self.path)
        self.pixmapWidget.setPixmap(pixmap)
        self.generateAndExportButton.setEnabled(True)
        self.setTitle()
        self.activateWindow()

    def fileChanged(self, path):
        #self.fileDropped(path)
        pass

    def transparentChanged(self):
        e = self.transparentCheckbox.isChecked()
        self.backgroundColorEdit.setEnabled(not e)
        self.backgroundColorLabel.setEnabled(not e)

    def forcePotChanged(self):
        e = self.forcePotCheckBox.isChecked()
        self.reorderTilesCheckBox.setEnabled(e)

    def generateAndExportClicked(self):

        g = Generator()
        g.tileWidth = self.tileWidthSpinBox.value()
        g.tileHeight = self.tileHeightSpinBox.value()
        g.forcePot = self.forcePotCheckBox.isChecked()
        g.isTransparent = self.transparentCheckbox.isChecked()
        g.bgColor = self.backgroundColorEdit.getColor()
        g.reorder = self.reorderTilesCheckBox.isChecked()
        g.padding = self.paddingSpinBox.value()

        target = g.create(self.pixmapWidget.pixmap);

        # export
        self.lastDir = os.path.dirname(self.path)
        targetPath = QFileDialog.getSaveFileName(self, 'Export', self.lastDir, 'PNG (*.png)')
        if targetPath:
            target.save(targetPath[0])
            showPixmap = QPixmap.fromImage(target)
            if self.showPixmapWidget:
                self.showPixmapWidget.deleteLater()
                del self.showPixmapWidget
            self.showPixmapWidget = PixmapWidget()
            self.showPixmapWidget.setWindowIcon(self.windowIcon())
            self.showPixmapWidget.setWindowTitle(os.path.basename(targetPath[0]))
            self.showPixmapWidget.resize(showPixmap.width(), showPixmap.height())
            self.showPixmapWidget.setPixmap(showPixmap)
            self.showPixmapWidget.show()

    def closeEvent(self, event):
        if self.showPixmapWidget:
            self.showPixmapWidget.close()

        # save settings
        self.settings.setValue('tileWidth', self.tileWidthSpinBox.value())
        self.settings.setValue('tileHeight', self.tileHeightSpinBox.value())
        self.settings.setValue('padding', self.paddingSpinBox.value())
        self.settings.setValue('forcePot', self.forcePotCheckBox.isChecked())
        self.settings.setValue('reorderTiles', self.reorderTilesCheckBox.isChecked())
        self.settings.setValue('transparent', self.transparentCheckbox.isChecked())
        self.settings.setValue('backgroundColor', self.backgroundColorEdit.getColor().name())
        self.settings.setValue('lastDir', self.lastDir)
        self.settings.setValue('MainWindow/geometry', self.saveGeometry())
        self.settings.setValue('MainWindow/windowState', self.saveState())

        super(MainWindow, self).closeEvent(event)

File: mainwindow.cpp
#include "mainwindow.h"

#include <QWidget>
#include <QHBoxLayout>
#include <QFileInfo>
#include <QDir>
#include <QFileDialog>
#include <QBuffer>
#include <QCoreApplication>
#include <QSettings>

MainWindow::MainWindow(QWidget *parent) : QMainWindow(parent) {
    QCoreApplication::setOrganizationName("Dynart");
    QCoreApplication::setOrganizationDomain("dynart.net");
    QCoreApplication::setApplicationName("TilePad");
    fileWatcher = new QFileSystemWatcher(this);
    connect(fileWatcher, &QFileSystemWatcher::fileChanged, this, &MainWindow::sourceFileChanged);
    createLayout();
    loadSettings();
    setWindowTitle("TilePad 0.5.0");
}

MainWindow::~MainWindow() {
}

QSize MainWindow::sizeHint() const {
    return QSize(800, 600);
}

void MainWindow::createLayout() {
    messageLabel = new QLabel("Error message");
    showInfo("Drop an image file to the grey area!");

    tileWidthSpinBox = new QSpinBox();
    tileWidthSpinBox->setFixedWidth(50);
    tileWidthSpinBox->setValue(16);

    tileHeightSpinBox = new QSpinBox();
    tileHeightSpinBox->setFixedWidth(50);
    tileHeightSpinBox->setValue(16);

    paddingSpinBox = new QSpinBox();
    paddingSpinBox->setFixedWidth(50);
    paddingSpinBox->setValue(1);

    forcePotCheckBox = new QCheckBox("Force PoT");
    forcePotCheckBox->setChecked(true);
    connect(forcePotCheckBox, &QCheckBox::checkStateChanged, this, &MainWindow::forcePotCheckBoxStateChanged);

    reorderCheckBox = new QCheckBox("Reorder tiles");

    removePaddingCheckBox = new QCheckBox("Remove padding");
    connect(removePaddingCheckBox, &QCheckBox::checkStateChanged, this, &MainWindow::removePaddingCheckBoxStateChanged);

    transparentCheckBox = new QCheckBox("Transparent");
    transparentCheckBox->setChecked(true);
    connect(transparentCheckBox, &QCheckBox::checkStateChanged, this, &MainWindow::transparentCheckBoxStateChanged);

    watchFileCheckBox = new QCheckBox("Watch file");
    connect(watchFileCheckBox, &QCheckBox::checkStateChanged, this, &MainWindow::watchFileCheckBoxStateChanged);

    backgroundColorEdit = new ColorEdit();
    backgroundColorEdit->setEnabled(false);

    sourcePixmapDropWidget = new PixmapDropWidget();
    sourcePixmapDropWidget->setMinimumHeight(300);
    sourcePixmapDropWidget->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    connect(sourcePixmapDropWidget, &PixmapDropWidget::dropSignal, this, &MainWindow::fileDropped);

    resultPixmapDropWidget = new PixmapDropWidget();
    resultPixmapDropWidget->setMinimumHeight(300);
    resultPixmapDropWidget->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    connect(resultPixmapDropWidget, &PixmapDropWidget::dropSignal, this, &MainWindow::fileDropped);

    tabWidget = new QTabWidget();
    tabWidget->addTab(sourcePixmapDropWidget, "Source");
    tabWidget->addTab(resultPixmapDropWidget, "Result");
    tabWidget->setStyleSheet("QTabWidget::pane { border: 0 solid white; } ");

    exportEdit = new QLineEdit();

    browseButton = new QPushButton("Browse ...");
    connect(browseButton, &QPushButton::clicked, this, &MainWindow::browseButtonClicked);

    reprocessButton = new QPushButton("Reprocess");
    reprocessButton->setEnabled(false);
    connect(reprocessButton, &QPushButton::clicked, this, &MainWindow::reprocess);

    exportButton = new QPushButton("Export");
    exportButton->setEnabled(false);
    connect(exportButton, &QPushButton::clicked, this, &MainWindow::exportButtonClicked);

    auto h = new QHBoxLayout();
    h->setContentsMargins(5, 5, 5, 5);
    h->addSpacing(5);
    h->addWidget(new QLabel("Tile width:"));
    h->addSpacing(5);
    h->addWidget(tileWidthSpinBox);
    h->addSpacing(15);
    h->addWidget(new QLabel("Tile height:"));
    h->addSpacing(5);
    h->addWidget(tileHeightSpinBox);
    h->addSpacing(15);
    h->addWidget(new QLabel("Padding:"));
    h->addSpacing(5);
    h->addWidget(paddingSpinBox);
    h->addSpacing(15);
    h->addWidget(forcePotCheckBox);
    h->addSpacing(15);
    h->addWidget(reorderCheckBox);
    h->addSpacing(15);
    h->addWidget(removePaddingCheckBox);
    h->addStretch();

    auto h2 = new QHBoxLayout();
    h2->setContentsMargins(5, 5, 5, 5);
    h2->addSpacing(5);
    h2->addWidget(transparentCheckBox);
    h2->addSpacing(15);
    h2->addWidget(new QLabel("Background color:"));
    h2->addSpacing(5);
    h2->addWidget(backgroundColorEdit);
    h2->addSpacing(15);
    h2->addWidget(watchFileCheckBox);
    h2->addStretch();

    auto h3 = new QHBoxLayout();
    h3->setContentsMargins(5, 5, 5, 5);
    h3->addWidget(exportEdit);
    h3->addSpacing(5);
    h3->addWidget(browseButton);
    h3->addSpacing(5);
    h3->addWidget(reprocessButton);
    h3->addSpacing(5);
    h3->addWidget(exportButton);

    auto centralWidget = new QWidget();
    auto v = new QVBoxLayout(centralWidget);
    v->setContentsMargins(0, 0, 0 ,0);
    v->setSpacing(0);
    v->addWidget(messageLabel);
    v->addLayout(h);
    v->addLayout(h2);
    v->addWidget(tabWidget);
    v->addLayout(h3);
    setCentralWidget(centralWidget);
}

void MainWindow::loadSettings() {
    QSettings settings;
    if (!settings.contains("tileWidth")) {
        return;
    }
    tileWidthSpinBox->setValue(settings.value("tileWidth").toInt());
    tileHeightSpinBox->setValue(settings.value("tileHeight").toInt());
    paddingSpinBox->setValue(settings.value("padding").toInt());
    forcePotCheckBox->setChecked(settings.value("forcePot").toBool());
    reorderCheckBox->setChecked(settings.value("reorder").toBool());
    removePaddingCheckBox->setChecked(settings.value("removePadding").toBool());
    transparentCheckBox->setChecked(settings.value("transparent").toBool());
    watchFileCheckBox->setChecked(settings.value("watchFile").toBool());
    backgroundColorEdit->setColorText(settings.value("backgroundColor").toString());
    exportEdit->setText(settings.value("exportPath").toString());
}

void MainWindow::closeEvent(QCloseEvent* event) {
    saveSettings();
    QMainWindow::closeEvent(event);
}

void MainWindow::saveSettings() {
    QSettings settings;
    settings.setValue("tileWidth", tileWidthSpinBox->value());
    settings.setValue("tileHeight", tileHeightSpinBox->value());
    settings.setValue("padding", paddingSpinBox->value());
    settings.setValue("forcePot", forcePotCheckBox->isChecked());
    settings.setValue("removePadding", removePaddingCheckBox->isChecked());
    settings.setValue("reorder", reorderCheckBox->isChecked());
    settings.setValue("transparent", transparentCheckBox->isChecked());
    settings.setValue("watchFile", watchFileCheckBox->isChecked());
    settings.setValue("backgroundColor", backgroundColorEdit->getColor().name());
    settings.setValue("exportPath", exportEdit->text());
}

void MainWindow::showError(QString text) {
    messageLabel->setStyleSheet("color: #fff; background: #a00; font-weight: bold; padding: 5px");
    messageLabel->setVisible(true);
    messageLabel->setText(text);
}

void MainWindow::showInfo(QString text) {
    messageLabel->setStyleSheet("color: #fff; background: #4286f4; font-weight: bold; padding: 5px");
    messageLabel->setVisible(true);
    messageLabel->setText(text);
}

void MainWindow::hideMessage() {
    messageLabel->setVisible(false);
}

void MainWindow::fileDropped(QString path) {
    hideMessage();
    if (!sourcePixmapDropWidget->load(path)) {
        showError("Couldn't load the image: " + path);
        return;
    }
    if (!currentSourcePath.isEmpty()) {
        fileWatcher->removePath(currentSourcePath);
    }
    currentSourcePath = path;
    if (watchFileCheckBox->isChecked()) {
        fileWatcher->addPath(currentSourcePath);
    }
    reprocessButton->setEnabled(true);
    reprocess();
    adjustUiAfterDrop(path);
}

void MainWindow::reprocess() {
    if (currentSourcePath.isEmpty()) {
        return;
    }
    hideMessage();
    QImage* image = createImageFromSource();
    QImage* resultImage;
    if (removePaddingCheckBox->isChecked()) {
        setUpRemover();
        resultImage = paddingRemover.create(image);
    } else {
        setUpGenerator();
        resultImage = paddingGenerator.create(image);
    }
    delete image;
    QPixmap* resultPixmap = new QPixmap(QPixmap::fromImage(*resultImage));
    resultPixmapDropWidget->setPixmap(resultPixmap);
    tabWidget->setCurrentIndex(1);
    update();
    showInfo("Reprocessing complete.");
}

void MainWindow::setUpGenerator() {
    paddingGenerator.setTileSize(tileWidthSpinBox->value(), tileHeightSpinBox->value());
    paddingGenerator.setPadding(paddingSpinBox->value());
    paddingGenerator.setForcePot(forcePotCheckBox->isChecked());
    paddingGenerator.setReorder(reorderCheckBox->isChecked());
    paddingGenerator.setTransparent(transparentCheckBox->isChecked());
    paddingGenerator.setBackgroundColor(backgroundColorEdit->getColor());
}

void MainWindow::setUpRemover() {
    paddingRemover.setTileSize(tileWidthSpinBox->value(), tileHeightSpinBox->value());
    paddingRemover.setPadding(paddingSpinBox->value());
}

QImage* MainWindow::createImageFromSource() {
    QPixmap* pixmap = sourcePixmapDropWidget->getPixmap();
    QByteArray bArray;
    QBuffer buffer(&bArray);
    buffer.open(QIODevice::WriteOnly);
    pixmap->save(&buffer, "PNG");
    buffer.close();
    QImage* image = new QImage(QImage::fromData(bArray));
    return image;
}

void MainWindow::adjustUiAfterDrop(QString path) {
    if (exportEdit->text().isEmpty()) {
        QFileInfo fileInfo(path);
        auto dir = fileInfo.absoluteDir().path();
        auto baseName = fileInfo.completeBaseName();
        auto suffix = fileInfo.suffix();
        exportEdit->setText(dir + "/" + baseName + ".export." + suffix);
    }
    exportButton->setEnabled(true);
}

void MainWindow::browseButtonClicked() {
    QFileDialog dialog(this);
    dialog.setFileMode(QFileDialog::AnyFile);
    dialog.setNameFilter(tr("Images (*.png *.jpg *.jpeg)"));
    if (dialog.exec()) {
        auto files = dialog.selectedFiles();
        if (files.length() > 0) {
            auto file = files.at(0);
            exportEdit->setText(file);
        }
    }
}

void MainWindow::transparentCheckBoxStateChanged(Qt::CheckState state) {
    backgroundColorEdit->setEnabled(state == Qt::Unchecked);
}

void MainWindow::forcePotCheckBoxStateChanged(Qt::CheckState state) {
    reorderCheckBox->setEnabled(state == Qt::Checked);
}

void MainWindow::removePaddingCheckBoxStateChanged(Qt::CheckState state) {
    reorderCheckBox->setEnabled(state == Qt::Unchecked && forcePotCheckBox->isChecked());
    forcePotCheckBox->setEnabled(state == Qt::Unchecked);
    transparentCheckBox->setEnabled(state == Qt::Unchecked);
    backgroundColorEdit->setEnabled(state == Qt::Unchecked && !transparentCheckBox->isChecked());
}

void MainWindow::watchFileCheckBoxStateChanged(Qt::CheckState state) {
    if (currentSourcePath.isEmpty()) {
        return;
    }
    if (state == Qt::Checked) {
        fileWatcher->addPath(currentSourcePath);
    } else {
        fileWatcher->removePath(currentSourcePath);
    }
}

void MainWindow::sourceFileChanged(const QString& path) {
    if (path != currentSourcePath || !watchFileCheckBox->isChecked()) {
        return;
    }
    if (!sourcePixmapDropWidget->load(currentSourcePath)) {
        return;
    }
    // Re-add the path since some systems remove it after a change
    fileWatcher->addPath(currentSourcePath);
    reprocess();
}

void MainWindow::exportButtonClicked() {
    auto exportPath = exportEdit->text();
    QFileInfo fileInfo(exportPath);
    auto format = fileInfo.suffix().toUpper();
    auto dir = fileInfo.dir();
    hideMessage();
    if (!dir.exists()) {
        showError("The export directory doesn't exist.");
        return;
    }
    if (format == "JPEG") {
        format = "JPG";
    }
    if (format != "PNG" && format != "JPG") {
        showError("The export extension must be .png, .jpg or .jpeg.");
        return;
    }
    QPixmap* pixmap = resultPixmapDropWidget->getPixmap();
    pixmap->save(exportPath, format.toStdString().c_str());
    showInfo("The export was successfull.");
}

